hazelcast/src/main/java/com/hazelcast/client/impl/clientside/ClusterDiscoveryServiceBuilder.java

```diff
@@ -132,22 +133,18 @@ private AddressProvider createAddressProvider(ClientConfig clientConfig, Discove
         isDiscoveryConfigurationConsistent(addressListProvided, awsDiscoveryEnabled, gcpDiscoveryEnabled, azureDiscoveryEnabled,
                 kubernetesDiscoveryEnabled, eurekaDiscoveryEnabled, discoverySpiEnabled, hazelcastCloudEnabled);
 
-        if (discoveryService != null) {
-            return new RemoteAddressProvider(() -> discoverAddresses(discoveryService), usePublicAddress(clientConfig));
-        } else if (hazelcastCloudEnabled) {
-            String discoveryToken;
-            if (cloudConfig.isEnabled()) {
-                discoveryToken = cloudConfig.getDiscoveryToken();
-            } else {
-                discoveryToken = cloudDiscoveryToken;
-            }
+        if (hazelcastCloudEnabled) {
+            String discoveryToken = cloudDiscoveryToken(cloudConfig, cloudDiscoveryToken);
             String cloudUrlBase = properties.getString(HazelcastCloudDiscovery.CLOUD_URL_BASE_PROPERTY);
             String urlEndpoint = HazelcastCloudDiscovery.createUrlEndpoint(cloudUrlBase, discoveryToken);
             int connectionTimeoutMillis = getConnectionTimeoutMillis(networkConfig);
             HazelcastCloudDiscovery cloudDiscovery = new HazelcastCloudDiscovery(urlEndpoint, connectionTimeoutMillis);
             return new RemoteAddressProvider(cloudDiscovery::discoverNodes, true);
+        } else if (!networkConfig.getAddresses().isEmpty()) {
+            return new DefaultAddressProvider(networkConfig);
```

Comment:
This branch seems unnecessary. Isn't this already covered on the last return statement?

Reply:
It was actually necessary because if there is any static address specified, it should take precedence over discovery strategies. But I refactored the condition. I think it's clearer now. PTAL.